### PR TITLE
Add initial Objective C support

### PR DESCRIPTION
**Pull request by scoopr**
Hi! I made an attempt to support parsing objective c headers. This is very much incomplete,
and likely not quite yet ready for merging, but I thought I'd share the progress so far.

Comments appreciated, I'm still very much a newbie in just about everything related to this change (rust as a language, rust ast, libclang, rust-bindgen), and there were many parts of code I wasn't quite sure about.

Commit message:
It parses interfaces and protocol but ignores base classes,
methods that don’t have arguments, the arguments are currently ignored.
Also you can pass objc class instances to C functions.

Next steps are inheritance/base classes, method signatures,
properties, categories.
Then check with system headers what is missing.

### CODE REVIEW

**Comment by scoopr** [comment]
The failure is due to me hardwiring the `use objc;` etc. for my testing.. I suppose the raw_line option is meant for that.

File: src/codegen/mod.rs [hunk 1]
```diff
@@ -2074,6 +2091,18 @@ impl ToRustTy for FunctionSig {
             // [1]: http://c0x.coding-guidelines.com/6.7.5.3.html
             let arg_ty = if let TypeKind::Array(t, _) = *arg_ty.canonical_type(ctx).kind() {
```
Comment (emilio) [inline src/codegen/mod.rs]: Instead of calling `canonical_type` two times, let's use a `match` expression.

File: src/codegen/mod.rs [hunk 2]
```diff
@@ -2189,6 +2218,91 @@ impl CodeGenerator for Function {
...
 }
 
+impl CodeGenerator for ObjCInterface {
+    type Extra = Item;
+    fn codegen<'a>(&self,
+                   ctx: &BindgenContext,
+                   result: &mut CodegenResult<'a>,
+                   _whitelisted_items: &ItemSet,
+                   _: &Item) {
+
```
Comment (emilio) [inline src/codegen/mod.rs]: nit: whitespace should go before attempting to merge.

File: src/codegen/mod.rs [hunk 3]
```diff
@@ -2189,6 +2218,91 @@ impl CodeGenerator for Function {
...
+
+
+        let mut impl_items = vec![];
+        let mut trait_items = vec![];
+
+        for method in &self.methods {
+            let method_name = ctx.rust_ident(&method.name);
+
+            let body = quote_stmt!(ctx.ext_cx(),msg_send![self, $method_name])
```
Comment (emilio) [inline src/codegen/mod.rs]: nit: space after `,`.

File: src/ir/comp.rs [hunk 4]
```diff
@@ -668,8 +668,11 @@ impl CompInfo {
 
                     // NB: This gets us an owned `Function`, not a
                     // `FunctionSig`.
-                    let signature = match Item::parse(cur, Some(potential_id), ctx) {
-                        Ok(item) if ctx.resolve_item(item).kind().is_function() => item,
+                    let signature =
+                        match Item::parse(cur, Some(potential_id), ctx) {
+                            Ok(item) if ctx.resolve_item(item)
```
Comment (emilio) [inline src/ir/comp.rs]: could you leave formatting patches in a different commit? That will make this much more pleasurable to review :)

File: src/ir/context.rs [hunk 5]
```diff
@@ -1070,6 +1071,7 @@ impl<'ctx> BindgenContext<'ctx> {
                 let name = item.canonical_path(self)[1..].join("::");
                 debug!("whitelisted_items: testing {:?}", name);
                 match *item.kind() {
+                    ItemKind::ObjCInterface(_) => true, // TODO:
```
Comment (emilio) [inline src/ir/context.rs]: Any particular reason you have a full `ItemKind` for this?

Does an interface have the same characteristics as, let's say, a `struct` in C? I mean, can you forward-declare them, point to them, etc?

If so, you probably want to just fold it into `Type` instead, since you'd share a lot of code for detecting duplicates, going to the definition of the type, etc.

(Edit: after reviewing the rest of the commit I've seen you've also added `TypeKind::ObjCInterface`, you might want to just use `ItemKind::Type` if there isn't a better reason to have a full `ItemKind`).

File: src/ir/objc.rs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+impl ClangSubItemParser for ObjCInterface {
+    fn parse(cursor: clang::Cursor,
+             _ctx: &mut BindgenContext)
+             -> Result<ParseResult<Self>, ParseError> {
+
+
+        match cursor.kind() {
+            CXCursor_ObjCInterfaceDecl |
+            CXCursor_ObjCProtocolDecl => {
```
Comment (emilio) [inline src/ir/objc.rs]: Hmmm, does libclang return a specific `CXTypeKind` for these (you can see this printing `cursor.cur_type()`)? If so, then merging it into `Type` is probably the best path forward, otherwise, it'll probably need a bit of thinking to get it right.

File: src/ir/ty.rs [hunk 7]
```diff
@@ -869,6 +874,60 @@ impl Type {
...
+            }
+            CXType_ObjCObjectPointer => {
+                //                panic!("yeah!");
+                println!("ty={:?}", ty);
+                println!("pointee_type={:?}", ty.pointee_type());
+
+                /*                let referenced = location.unwrap().referenced().unwrap();
+                let referenced_ty = referenced.cur_type();
+                let declaration = referenced_ty.declaration();
+                println!("referenced={:?}", referenced);
```
Comment (emilio) [inline src/ir/ty.rs]: Heh, welcome to the joy of debugging libclang :)

File: src/ir/ty.rs [hunk 8]
```diff
@@ -869,6 +874,60 @@ impl Type {
...
+                                     potential_id,
+                                     ctx);
+                println!("param named_type={:?}", param);
+                println!("element type ={:?}", ty.elem_type());
+                println!("typedef type ={:?}",
+                         location.unwrap().typedef_type());
+                println!("cur type ={:?}", location.unwrap().cur_type());
+                println!("objc encoding ={:?}", ty.objc_encoding());*/
+
+                let inner = Item::from_ty_or_ref(ty.pointee_type().unwrap(),
```
Comment (emilio) [inline src/ir/ty.rs]: This pretty much shares the body with `CXType_Pointer`, reference, et. al, right? If so, let's just fold it there when you're done with debugging it :)

File: src/ir/ty.rs [hunk 9]
```diff
@@ -869,6 +874,60 @@ impl Type {
                                            parent_id,
                                            ctx);
             }
+            CXType_ObjCInterface => {
```
Comment (emilio) [inline src/ir/ty.rs]: Yeah, probably you don't want a new `itemid` for each Interface you parse, and I don't see any deduplication going on, so probably you should make this the actual container of your data (and use `cursor.definition()` to properly go to the actual definition, just in case you find a forward declaration before). You probably need to do roughly the same thing we do for `CXType_Record` (which is a plain struct), just storing most of the specific interface-related data you want.

File: src/lib.rs [hunk 10]
```diff
@@ -697,6 +697,12 @@ impl<'ctx> Bindings<'ctx> {
             try!(writer.write("\n".as_bytes()));
         }
 
+        // TODO:
+
+        try!(writer.write("use objc;\n\n".as_bytes()));
```
Comment (emilio) [inline src/lib.rs]: Instead of printing this here we should use something like what we do to insert the `__BindgenUnionField`, etc. See `CodegenResult::saw_union` and similar methods.

File: src/ir/comp.rs (hunk 4 above)
Comment (scoopr) [inline src/ir/comp.rs]: oops, I thought I cleaned them all. I have the editor set up on formatting on save, and I manually cleaned up rest of these.

File: src/ir/context.rs (hunk 5 above)
Comment (scoopr) [inline src/ir/context.rs]: Frankly, I wasn't sure, and I think I just looked at the try_parse! and thought ‘thats what I need’.

Forward declaration is a bit weird, using `@class Foo;`, this gives a cursor type CXCursor_ObjCClassRef, have not looked at that more closely.

One other interesting bit is protocols, those are for extending more methods for existing interface. Those come in as CXCursor_ObjCCategoryDecl, so I guess they could be handled completely separate (and generate their own trait)

File: src/ir/objc.rs (hunk 6 above)
Comment (scoopr) [inline src/ir/objc.rs]: It does, so I did that.

One annoying hack was that protocols were not types, so I forced it to think they are interfaces.

File: src/ir/ty.rs (hunk 7 above)
Comment (scoopr) [inline src/ir/ty.rs]: got me with my pants down :)

File: src/ir/ty.rs (hunk 8 above)
Comment (scoopr) [inline src/ir/ty.rs]: done

File: src/lib.rs (hunk 10 above)
Comment (scoopr) [inline src/lib.rs]: done

**Comment by scoopr** [comment]
I'm still missing documentation and testing.
How should I make it run some of the headers in -x objective-c? I suppose I could just name them with .m suffix, or add somehow add the -x flag when header_name.containts("objc")

**Comment by emilio** [comment]
Right now we rely on the automatic language detection for most of the tests, but you can annotate a test with `// bindgen-flags:`, and they'll get passed to the harness.

So it seems you either want `m` files (that's fine by me), or otherwise do something like what we do in `tests/headers/dash_language.h`, which does `// bindgen-flags: -- -x c++`.

File: src/codegen/mod.rs [hunk 11]
```diff
@@ -2095,10 +2117,27 @@ impl ToRustTy for FunctionSig {
             //     the array type derivation.
             //
             // [1]: http://c0x.coding-guidelines.com/6.7.5.3.html
-            let arg_ty = if let TypeKind::Array(t, _) = *arg_ty.canonical_type(ctx).kind() {
+
+
+            let arg_ty = match *arg_ty.canonical_type(ctx).kind()
```
Comment (emilio) [inline src/codegen/mod.rs]: nit: indentation and spacing is quite odd in all this block right now, not sure if intended :)

File: src/codegen/mod.rs [hunk 12]
```diff
@@ -2062,6 +2077,13 @@ impl ToRustTy for Type {
                 let ident = ctx.rust_ident(&name);
                 quote_ty!(ctx.ext_cx(), $ident)
             }
+
+            TypeKind::ObjCInterface(..) => {
+                let name = item.canonical_name(ctx);
```
Comment (emilio) [inline src/codegen/mod.rs]: This will be enough if you can't mix objective-c with namespaces. If you can, you need to handle them somehow (hint: you could use `utils::build_templated_path`, but pass always an empty `vec![]` of template arguments).

File: src/ir/objc.rs (hunk 6 above)
Comment (emilio) [inline src/ir/objc.rs]: What kind of type do they report?

File: src/codegen/mod.rs (hunk 12 above)
Comment (scoopr) [inline src/codegen/mod.rs]: clang will give an actual error if interface embedded in namespace. And to even get that error, it needs to be run with -x objective-c++ mode

File: src/codegen/mod.rs (hunk 11 above)
Comment (scoopr) [inline src/codegen/mod.rs]: huh, I wonder how atom/rustfmt missed that o_O

**Comment by scoopr** [comment]
Heres my latest, with some rudimentary tests and docs. 
Also fixed some objc pointers in structs and global variables to emit the 'id' type there as well.

I'm using the id here to mimic the cocoa-rs convention, thought I think some have suggested trying `struct Foo(id)` type of wrapping of the type, and impl the traits on those types. But I think that kind of experimentation is ok to day later when it otherwise works.

**Comment by emilio** [comment]
You need to add the `objc` dependency to the `tests/expectations` directory, so they compile.

**Comment by scoopr** [comment]
Quite, it didn't seem quite as simple.
Either I did something wrong that I don't understand, or things really are this complicated, but I ended up adding a bindgen flag, that toggles the generated prologue from emitting either `use objc;` or `#[macro_use] extern crate objc;`, as the generated tests were run individually, and adding the extern crate to the tests lib.rs didn't seem work.

**Comment by scoopr** [comment]
Ehh, botched rebase, I was sure I had run the tests one more time. Sorry for the noise

**Comment by scoopr** [comment]
ngggh. of course, it needs some cfg guards

**Comment by scoopr** [comment]
Woo, finally a passing commit! :D

I opted to add cfg(target_os="macos") to the objc tests. It seemed to be the simplest change, though I'm not sure the .contains("objc_") is good enough toggle for it. I guess this also means that the travis config would need a osx target as well, to have those tests be exercised.

File: tests/tests.rs [hunk 13]
```diff
@@ -98,10 +98,17 @@ fn create_bindgen_builder(header: &PathBuf)
     let header_str = try!(header.to_str()
         .ok_or(Error::new(ErrorKind::Other, "Invalid header file name")));
 
+    let objc_guard;
+    if header_str.contains("objc_") {
```
Comment (emilio) [inline tests/tests.rs]: instead of this, let's add a new line to the test configuration (see the `if line.contains(bindgen-flags` check above). `// bindgen-osx-only` maybe?

File: src/ir/objc.rs [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+            CXChildVisit_Continue
+        });
+        Some(interface)
+    }
+}
+
+
+impl ObjCInstanceMethod {
+    fn new(name: &str) -> ObjCInstanceMethod {
+
```
Comment (emilio) [inline src/ir/objc.rs]: please remove this initial newline.

File: src/ir/objc.rs [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+        let split_name: Vec<&str> = name.split(':').collect();
+
```
Comment (emilio) [inline src/ir/objc.rs]: This may also go away if you want :)

File: src/ir/objc.rs [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+    /// List of the methods defined in this interfae
+    pub methods: Vec<ObjCInstanceMethod>,
+}
+
+/// The objective c methods
+#[derive(Debug)]
+pub struct ObjCInstanceMethod {
+    /// The original method selector name
+    /// like, dataWithBytes:length:
+    pub name: String,
```
Comment (emilio) [inline src/ir/objc.rs]: instead of making these public, could/should we make them private and add accessors?

File: src/ir/objc.rs [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+///
+/// Also protocols are parsed as this type
+#[derive(Debug)]
+pub struct ObjCInterface {
+    /// The name
+    /// like, NSObject
+    pub name: String,
+
+    /// List of the methods defined in this interfae
+    pub methods: Vec<ObjCInstanceMethod>,
```
Comment (emilio) [inline src/ir/objc.rs]: instead of making these public, could/should we make them private and add accessors?

File: src/codegen/mod.rs [hunk 18]
```diff
@@ -2294,6 +2412,38 @@ mod utils {
     use syntax::ast;
     use syntax::ptr::P;
 
+
+    pub fn prepend_objc_header(ctx: &BindgenContext,
+                               result: &mut Vec<P<ast::Item>>) {
+
+
```
Comment (emilio) [inline src/codegen/mod.rs]: remove these two newlines

File: src/codegen/mod.rs [hunk 19]
```diff
@@ -2294,6 +2412,38 @@ mod utils {
...
+        let use_objc;
+
+        if ctx.options().objc_extern_crate {
```
Comment (emilio) [inline src/codegen/mod.rs]: why not:

```rust
let use_objc = if ctx.options().objc_extern_crate {
    quote_item!(...)
} else {
    quote_item!(...)
};

File: src/codegen/mod.rs (hunk 2 above)
Comment (emilio) [inline src/codegen/mod.rs]: Please remove this newline

File: src/codegen/mod.rs [hunk 20]
```diff
@@ -2261,6 +2295,90 @@ impl CodeGenerator for Function {
...
+                rules: ast::BlockCheckMode::Default,
+                span: ctx.span(),
+            };
+
+            let sig = aster::AstBuilder::new()
+                .method_sig()
+                .unsafe_()
+                .fn_decl()
+                .self_()
+                .build(ast::SelfKind::Value(ast::Mutability::Immutable))//ast::SelfKind::Region(None, ast::Mutability::Mutable))
```
Comment (emilio) [inline src/codegen/mod.rs]: remove or reword the comment please :)

File: src/codegen/mod.rs [hunk 21]
```diff
@@ -2142,10 +2162,24 @@ impl ToRustTy for FunctionSig {
             //     the array type derivation.
             //
             // [1]: http://c0x.coding-guidelines.com/6.7.5.3.html
-            let arg_ty = if let TypeKind::Array(t, _) = *arg_ty.canonical_type(ctx).kind() {
+
+
+            let arg_ty = match *arg_ty.canonical_type(ctx).kind() {
```
Comment (emilio) [inline src/codegen/mod.rs]: please fix the alignment in this block.

File: src/codegen/mod.rs [hunk 22]
```diff
@@ -2142,10 +2162,24 @@ impl ToRustTy for FunctionSig {
             //     the array type derivation.
             //
             // [1]: http://c0x.coding-guidelines.com/6.7.5.3.html
-            let arg_ty = if let TypeKind::Array(t, _) = *arg_ty.canonical_type(ctx).kind() {
+
```
Comment (emilio) [inline src/codegen/mod.rs]: please remove the whitespace between the comment and the `let`.

File: src/codegen/mod.rs [hunk 23]
```diff
@@ -2261,6 +2295,90 @@ impl CodeGenerator for Function {
...
+
+        let trait_block = aster::AstBuilder::new()
+            .item()
+            .pub_()
+            .trait_(&self.name)
+            .with_items(trait_items)
+            .build();
+
+        let ty_for_impl = aster::AstBuilder::new()
```
Comment (emilio) [inline src/codegen/mod.rs]: this is the same as: `quote_ty!(ctx.ext_cx(), id)`, which I think is clearer.

File: src/codegen/mod.rs [hunk 24]
```diff
@@ -2109,6 +2125,10 @@ impl ToRustTy for Type {
                 let ident = ctx.rust_ident(&name);
                 quote_ty!(ctx.ext_cx(), $ident)
             }
+            TypeKind::ObjCInterface(..) => {
+                let ident = ctx.rust_ident("id");
+                quote_ty!(ctx.ext_cx(), $ident)
```
Comment (emilio) [inline src/codegen/mod.rs]: just `quote_ty!(ctx.ext_cx(), id)` should work.

File: tests/tests.rs (hunk 13 above)
Comment (scoopr) [inline tests/tests.rs]: Ooh, that makes so much more sense, obvious even.

File: src/ir/objc.rs (hunk 17 above)
Comment (scoopr) [inline src/ir/objc.rs]: Sure

File: src/codegen/mod.rs (hunk 24 above)
Comment (scoopr) [inline src/codegen/mod.rs]: Ah didn't realise that, lot cleaner for sure!

File: src/codegen/mod.rs (hunk 21 above)
Comment (scoopr) [inline src/codegen/mod.rs]: Damnit, I fixed this once already, but somehow rustfmt got really confused in this block :(

File: src/codegen/mod.rs (hunk 19 above)
Comment (scoopr) [inline src/codegen/mod.rs]: At first I wrote it like that, and I now put it back like that. Somehow the style doesn't quite appeal to me, especially if the condition is long and/or the blocks are sprawling. I think I got it decent now with rustfmt disabled, which made the unwrap hanging awkwardly.

File: tests/tests.rs [hunk 25]
```diff
@@ -87,6 +87,12 @@ fn create_bindgen_builder(header: &PathBuf)
         } else if line.contains("bindgen-unstable") &&
                   cfg!(feature = "llvm_stable") {
             return Ok(None);
+        } else if line.contains("bindgen-osx-only") {
```
Comment (emilio) [inline tests/tests.rs]: You need to use `take(3)` above now :)

File: src/codegen/mod.rs [hunk 26]
```diff
@@ -2144,10 +2163,22 @@ impl ToRustTy for FunctionSig {
...
+            let arg_ty = match *arg_ty.canonical_type(ctx).kind() {
+                TypeKind::Array(t, _) => {
                 t.to_rust_ty(ctx).to_ptr(arg_ty.is_const(), ctx.span())
+                },
+                TypeKind::Pointer(inner) => {
+                    let inner = ctx.resolve_item(inner);
+                    let inner_ty = inner.expect_type();
+                    if let TypeKind::ObjCInterface(_) = *inner_ty.canonical_type(ctx).kind() {
+                        aster::ty::TyBuilder::new().id("id")
             } else {
```
Comment (emilio) [inline src/codegen/mod.rs]: Wow, I guess rustfmt really likes to mess with this assignment?

Comment (scoopr) [inline src/codegen/mod.rs]: that or I forgot add the chunk, nevertheless, I've disabled rustfmt, and hopefully it now makes sense

File: tests/tests.rs (hunk 25 above)
Comment (scoopr) [inline tests/tests.rs]: Fine :)

**Comment by emilio** [comment]
Looks good to me! I've opened https://github.com/servo/rust-bindgen/pull/464 to re-enable testing on OSX, after that lands this is good to go \o/

**Comment by emilio** [comment]
@bors-servo r+

Thanks for this! :)
